README.md
# covid-data-analysis

[![GitHub commit](https://img.shields.io/github/last-commit/gerrygeko/covid-data-analysis)](https://github.com/gerrygeko/covid-data-analysis)[![License: GPL v2](https://img.shields.io/badge/License-GPL%20v2-blue.svg)](https://www.gnu.org/licenses/old-licenses/gpl-2.0.en.html)[![Python 3.10](https://img.shields.io/badge/python-3.10-blue.svg)](https://www.python.org/)

This web application gathers and shows World data for the COVID-19, with particular attention to the details of the 
Italian regions. 
It gives the possibility to arrange and visualize the data in different ways using different graphs.
The web application is reachable at the following link [Coronavirus Data-Gellex](https://www.data-covid.com/)

## Data update
- Italian Data (data of the day) is published every day from the Department of Civil Protection at about 17:30.
- Worldwide Data (data of the previous day) is released by Data Packaged Core Datasets at about 08:00
- Update to this application is expected slightly after, but to keep consistency in case of later updates, 
every 30 minutes a new attempt for check the presence of new data will be made

## Technology adopted
- Programming language adopted for the whole project is [Python](https://www.python.org/)
- The framework used to build this web-based analytic application is [Plotly - Dash](https://plotly.com/dash/)
- The web application is running a cloud solution via the application platform [Heroku](https://www.heroku.com/) 
(PaaS - Platform as a Service)

## Changelog
- Coronavirus Data-Gellex updates can be followed [here](https://github.com/gerrygeko/covid-data-analysis/blob/master/CHANGELOG.md)

## Credits
- Italian data is offered by [Department of Civil Protection](https://github.com/pcm-dpc/COVID-19).
- Worldwide data is offered by [Johns Hopkins University Center for Systems Science and Engineering ](https://github.com/CSSEGISandData) and normalized by [DataHub.io](https://github.com/datasets/covid-19)
- The news feed is offered by [News API](https://newsapi.org/)
- Resident Population data is offered by [ISTAT](http://demo.istat.it/popres/index.php?anno=2020&lingua=ita)
- Italian Intensive Care data is provided by [ilsole24ore](https://www.ilsole24ore.com/art/coronavirus-terapie-intensive-aumento-quali-regioni-sono-pronte-la-seconda-ondata-ADNUkdv)

## Contact Us
Gerardo Lamorte
- Email: lamorte.gerardo@gmail.com
- [LinkedIn](https://www.linkedin.com/in/gerardo-lamorte-a25928149/) profile

Nicola Mastrangelo
- Email: mastrangelo.nicola@gmail.com
- [LinkedIn](https://www.linkedin.com/in/nicola-mastrangelo-240810107/) profile


CHANGELOG.md
# 4.0.11
#### Updated
* Data update according to the new standards of data providers

# 4.0.10
#### Updated
* Introduce performance improvements that speed up the loading of the page

#### Changed
* Http traffic is always redirect to Https from now on

# 4.0.9
#### Updated
* Python has been upgraded from version 3.7 to version 3.10 bringing in updates and security fixes

#### Changed
* Changed name of the application from "SARS-CoV-2-Gellex" to "Coronavirus Data" to describe better the purpose of the website

#### Added
* Added support for MacOS, Android and Windows platforms. Option to save the link from the browser to your home screen, now gives an app-like experience when accessed
* It is possible to check the number of unique visitors to the platform. The new text is located at the footer of the webpage

# 4.0.8
#### Fixed
* The chart of administrations by age is back online
#### Changed
* Changed the calculation of the target of 90% of possible vaccinations. 
The number of days needed to reach the desired information, in line with current government target, is made by the division between the number of doses remaining to be administered and 
the average number (7-day rolling average) of first, second, single-dose, and prior infections each day
#### Updated
* Updated the information of total of vaccinated people and percentage cards
#### Added
* Added more infos on administration chart about administrated vaccines grouped by suppliers

# 4.0.7
#### Added
* Additional doses group chart (grouped by age)
#### Changed
* Calculation of the series shown in the administrated vaccines chart
* Target of Italian vaccination campain is now focused on 90% of vaccinable population 
*N.B.The vaccinable Italian population is represented by the over 12s. The audience (~ 54 million people) is based on 
the data of the holders of health cards or health certificates such as the "foreigners temporarily present" card 
recognized to irregular migrants*

# 4.0.6
#### Added
* "Buy Me a coffee" button- Support the project :)

# 4.0.5
#### Changed
* Updated the italian population according to the latest ISTAT updates

# 4.0.4
#### Fixed
* Removed field "categories" from vaccines charts since is not contained anymore in the open data used for this project

# 4.0.3
#### Changed
* Updated categories showed in the vaccines charts (grouped by age)

#### Fixed
* Fixed a bug showing an incorrect herd immunity prediction

# 4.0.2
#### Changed
* Changed the strategy to predict Herd Immunity Date; now the calculation takes into account the daily administrations and not exclusively the number of second doses

# 4.0.1
#### Fixed
* Fixed a bug in the italian cards which not show the correct color control for Pressure ICU and Positive/Swabs Ratio

#### Added
* "#datiBeneComune" references in the Credits section
* Copyright

# 4.0.0
#### Added
* Italy Vaccines Tab: the web application now shows italian data about Vaccination Campaign report.
    * Data Cards Italy: administered doses, delivered doses, total people vaccinated, percentage vaccinated population
    * Daily vaccinated people (presumed date of attainment of herd immunity)
    * Daily administrations
    * Administrations by age
    * Regional Data

# 3.0.6
#### Fixed
* Update of URL World Data repository to make a correct loading of the data starting from January 1, 2021

# 3.0.5
#### Fixed
* Replaced "googletrans" lib with "pygoogletranslation" to resolve issues caused by some changes on the 
  Google Translation API

# 3.0.4
#### Added
* Introduced card that shows the pressure in the ICU also in the regional tab

#### Fixed
* Fixed a bug in the translation library that was preventing the website to be available

# 3.0.3
#### Added
* Italy Tab: 2 new cards are now available:
    * ratio new positives / new swabs
    * ICU pressure

# 3.0.2
#### Fixed
* Fixed a bug that in specific cases, did not perform a correct association of nations to available population

# 3.0.1
#### Added
* DataHub references in the Credits section.

#### Changed
* Last update label now shows also the source of data

# 3.0.0
#### Added
* Worldwide Tab: the web application now shows worldwide data from all 188 countries that release Covid-19 related reports
    the following components show the updated data through the selection of the data of interest to the user:
    * Total Data Cards: show the official total world data;
    * Data Line Chart: shows the trend of the selected data;
    * Data Table: shows, in order of value importance, the ranking of countries;
    * Multiple Line Chart: allows you to compare the trend of the selected data from different countries;
    * Country Data Cards: show the official data released by the various countries of the world;
#### Changed 
* Italy Tab is now in line with World Tab layout

# 2.0.2
#### Changed
* Improvement of application layout
* Update of Italian Population Data by ISTAT (last official update 01/01/2020)

# 2.0.1
#### Fixed
* Fixed a bug that in specific cases, did not perform a correct translation

# 2.0.0
#### Added
* Localization is now available and the following languages are supported:
    * Italian
    * English
    * French
    * German
    * Spanish
    * Dutch
    * Portoguese 
#### Fixed
* Improved (mobile) navigation while scrolling through the different graphs 

# 1.1.2
#### Changed
* Display details of regions in the bar graphs with better format
* Hovering on any point of the main graph, will display information collected in the same window 
for all the regions selected instead of just displaying information of a single region
#### Added
* ICU and Swabs National Data Cards
#### Fixed
* Updates for new data showed in the graphs is more reliable

# 1.1.1
#### Changed
* National Data Cards has been removed from tabs component and the layout has been optimized for mobile view
* Italian map now shows the different brackets in different shades of blue
#### Fixed

resources.py
from pygoogletranslation import Translator

import logger
import time
from constants import DEFAULT_LANGUAGE

language_list = [{'label': 'Italiano', 'value': 'it'},
                 {'label': 'English', 'value': 'en'},
                 {'label': 'Dutch', 'value': 'nl'},
                 {'label': 'German', 'value': 'de'},
                 {'label': 'French', 'value': 'fr'},
                 {'label': 'Spanish', 'value': 'es'},
                 {'label': 'Portuguese', 'value': 'pt'}
                 ]

standard_colors = dict(light_blue="rgb(123, 199, 255)",
                       dark_blue="rgb(4, 74, 152)",
                       pink="rgb(244, 0, 161)",
                       aqua="rgb(0, 204, 153)",
                       azure="rgb(51, 133, 255)",
                       red="rgb(204, 51, 0)")


class LocaleLanguage:

    def __init__(self, language):
        self.language = language

    def __str__(self):
        return f"ID: {id(self)} and language {self.language}"


resources = {DEFAULT_LANGUAGE: {
    'denominazione_regione': 'Regione',
    'ricoverati_con_sintomi': 'Ricoverati con sintomi',
    'terapia_intensiva': 'Terapia intensiva',
    'totale_ospedalizzati': 'Totale ospedalizzati',
    'isolamento_domiciliare': 'Isolamento domiciliare',
    'totale_positivi': 'Casi attivi',
    'variazione_totale_positivi': 'Variazione totale positivi',
    'nuovi_positivi': 'Nuovi positivi',
    'dimessi_guariti': 'Guariti',
    'deceduti': 'Decessi',
    'casi_da_sospetto_diagnostico': 'Casi Positivi identificati dal Sospetto Diagnostico',
    'casi_da_screening': 'Casi Positivi accertati da Screening',
    'totale_casi': 'Totale casi',
    'tamponi': 'Tamponi effettuati',
    'casi_testati': 'Casi Testati',
    'ratio_n_pos_tamponi': 'Rapporto Positivi / Tamponi',
    'pressure_ICU': 'Pressione terapie intensive',
    'nazione': 'Italia',
    'label_dati_italia': 'Totale Dati - Italia',
    'label_dati_mondo': 'Totale Dati - Mondo',
    'label_tab_1': 'Analisi dei dati nazionali',
    'label_tab_2': 'Analisi dei dati regionali',
    'label_tab_master_worldwide': 'Mondo',
    'label_tab_master_ita': 'Italia',
    'label_tab_master_ita_regions': 'Regioni Italiane',
    'label_tab_vaccines_italy': 'Tassi di vaccinazione Italia',
    'label_select_data': 'Seleziona il dato da analizzare:',
    'label_select_multiregion': 'Seleziona una o più regioni italiane da confrontare:',
    'label_select_multicountry': 'Seleziona uno o più Paesi da confrontare:',
    'label_select_region': 'Seleziona la regione italiana da analizzare:',
    'label_select_country': 'Paese:',
    'label_data_picco_max': 'Data Picco massimo',
    'label_valore_picco_max': 'Picco massimo',
    'label_media_contagi': 'Media contagi',
    'label_casi_attivi': 'Casi Attivi',
    'label_news': 'News Scienza - Italia',
    'label_contact_us': 'Contattaci',
    'header_last_update_world': 'Ultimo rilascio dati DataHub:',
    'header_last_update_italy': 'Ultimo rilascio dati Protezione Civile Italiana:',
    'header_last_update_vaccines_italy': 'Ultimo rilascio dati Ministero della Salute:',
    'label_last_news_update': 'Ultimo aggiornamento: ',
    'label_last_check_update': 'Ultimo controllo dati:',
    'label_map': 'N° casi ogni 100K abitanti',
    'label_top_ten': 'Top 10 Regioni:',
    'label_hover_map': 'per regione (ogni cento mila abitanti)',
    'label_hover_world_map': 'per nazione (ogni cento mila abitanti)',
    'label_persone': 'Persone',
    'vaccine_phase_1': 'Fase 1',
    'vaccine_phase_2': 'Fase 2',
    'vaccine_phase_3': 'Fase 3',
    'vaccine_phase_4': 'Fase 4',
    'label_credits': 'Crediti',
    'confirmed_worldwide_aggregate': 'Totale Casi',
    'recovered_worldwide_aggregate': 'Guariti',
    'deaths_worldwide_aggregate': 'Decessi',
    'increase_rate_worldwide_aggregate': 'Tasso di Crescita',
    'Country': 'Paese',
    'Confirmed': 'Confermati',
    'Recovered': 'Guariti',
    'Deaths': 'Decessi',
    'New Confirmed': 'Nuovi Positivi',
    'New Recovered': 'Nuovi Guariti',
    'New Deaths': 'Nuovi Decessi',
    'Active_cases': 'Casi Attivi',
    'total_administered_doses': 'Dosi totali somministrate',
    'administered_doses': 'Dosi somministrate',
    'delivered_doses': 'Dosi Consegnate',
    'total_people_vaccinated': 'Totale Persone Vaccinate (over 5)',
    'percentage_vaccinated_population': 'Percentuale Platea Vaccinati (over 5)',
    'administrations_by_age': 'Somministrazioni per fasce di età',
    'administrations_by_day': 'Somministrazioni giornaliere',
    'total_administrations': 'Somministrazioni totali',
    'ratio_percentage_administrations': 'Rapporto % somministrazioni',
    'filter_by': 'Filtra per:',
    'total_vaccines': 'Totale',
    'suppliers': 'Fornitori',
    'sex': 'Sesso',
    'doses': 'Dosi',
    'doses_first_second': 'Prima e seconda dose',
    'additional_doses': 'Dosi addizionali',
    'sex_male': 'Maschi',
    'sex_female': 'Femmine',
    'age_60_69': 'Età 60-69',
    'age_70_79': 'Età 70-79',
    'over_80': 'Over 80',
    'first_vaccine_dose': 'Prima dose',
    'second_vaccine_dose': 'Seconda dose',
    'previous_infection_vaccine_dose': 'Dose unica per pregressa infezione',
    'additional_vaccine_dose': 'Dose aggiuntiva',
    'booster_vaccine_dose': 'Dose richiamo',
    'daily_administrations': ' Somministrazioni',
    'immunization_rate_label': 'Ritmo di immunizzazione',
    'rolling_average_7_days': 'Media mobile a 7 giorni',
    'string_italy_herd_immunity': 'Data raggiungimento 90% popolazione vaccinabile: ',
    'unique_visitors': 'Visitatori unici: ',
    'test_update': 'Test di notifica',
    'notification_text_repo_pandemic_data_ita_updated': '"Nuovi dati giornalieri dalla Protezione Civile Italiana, '
                                                        'controlla gli aggiornamenti!"',
    'notification_text_repo_pandemic_data_world_updated': 'Nuovi dati mondiali sui contagi Covid-19, controlla gli '
                                                          'aggiornamenti!',
    'notification_text_repo_vaccines_data_ita_updated': 'Nuovi dati sulla situazione vaccini in Italia, controlla gli '
                                                        'aggiornamenti! ',
    'notification_text_daily_italian_update': "Bollettino giornalerio del {}: {} casi covid in Italia!"
}
}

log = logger.get_logger()
locale_language = LocaleLanguage(DEFAULT_LANGUAGE)


def load_resource(name):
    try:
        resource = find_resource(name)
    except KeyError:
        resource = "Label not available"
    return resource


def find_resource(name):
    if locale_language.language not in resources.keys():
        raise KeyError(
            f"The language selected is not loaded: {locale_language.language}. Maybe you are running in debug mode ")
    if name not in resources[locale_language.language].keys():
        raise KeyError(f"The resource you are trying to load is not present. Resource: {name}")
    return resources[locale_language.language][name]


def create_resource_dict_for_languages(resource_key, is_debug, *args):
    if is_debug:
        # If in debug mode we pass the italian resource as an english resource,
        # because the OneSignal API require 'en' to be default, but in debug we have only 'IT' resources.
        return {"en": resources[DEFAULT_LANGUAGE][resource_key].format(*args)}
    else:
        languages = [language_dict['value'] for language_dict in language_list]
    resource_dict = {language.lower(): resources[language][resource_key].format(*args) for language in languages}
    return resource_dict


def add_translations_to_resource_dict(translations, language):
    resources_key_list = list(resources[DEFAULT_LANGUAGE].keys())
    language_dict = {language: {}}
    resources.update(language_dict)
    for key, translation in zip(resources_key_list, translations):
        resources[language][key] = translation.text


def start_translation():
    translator = Translator()
    languages_code_list = []
    log.info(f"Start of translation service")
    for language in language_list:
        languages_code_list.append(language["value"])

    default_language_string_list = list(resources[DEFAULT_LANGUAGE].values())

    for language in languages_code_list:
        if language != DEFAULT_LANGUAGE:
            start_time = time.time()
            translations = translator.translate(default_language_string_list, src=DEFAULT_LANGUAGE, dest=language)
            add_translations_to_resource_dict(translations, language)
            log.info(f"The {language} translation has taken {time.time() - start_time} seconds")
            log.info(f"End of loading for resources for: {language}")

    log.info(f"Translation for all languages is completed")
